pkg/kubelet/container/interface.go

```diff
@@ -0,0 +1,296 @@
+	State SandboxState
+}
+
+// Sandbox provides isolation with resource requirements in which containers
+// can be run.
+type Sandbox struct {
+	// ID of the sandbox.
+	ID string
+	// Config of the sandbox used to create the sandbox.
+	Config SandboxConfig
+	// Status of the sandbox.
+	Status SandboxStatus
+}
+
+// SandboxStatus contains the status of the sandbox.
+type SandboxStatus struct {
+	// State of the sandbox.
+	State SandboxState
+	// Network contains network status if network is handled by the runtime.
+	Network *SandboxNetworkStatus
+	// Status specific to a Linux sandbox.
+	Linux *LinuxSandboxStatus
+}
+
+type SandboxNetworkStatus struct {
+	IP string
+}
+
+// Namespaces contains paths to the namespaces.
+type Namespaces struct {
+	// Network is the path to the network namespace.
+	Network string
+}
+
+// LinuxSandBoxStatus contains status specific to Linux sandboxes.
+type LinuxSandboxStatus struct {
+	// Namespaces contains paths to the sandbox's namespaces.
+	Namespaces *Namespaces
+}
+
+// Resources contains the CPU/memory resource requirements.
+type Resources struct {
+	// CPU resource requirement.
+	CPU api.ResourceRequirements
+	// Memory resource requirement.
+	Memory api.ResourceRequirements
+}
+
+// ContainerRuntime provides methods for container lifecycle operations, as
+// well as listing or inspecting existing containers.
+type ContainerRuntime interface {
+	// Create creates a container in the sandbox, and returns the ID of hte
+	// created container.
+	Create(config *ContainerConfig, sandboxConfig *SandboxConfig, sandboxID string) (string, error)
+	// Start starts a created container.
+	Start(id string) error
+	// Stop stops a running container with a grace period (i.e., timeout).
+	Stop(id string, timeout int) error
+	// Remove removes the container.
+	Remove(id string) error
+	// List lists the existing containers that matches the ContainerFilter.
+	List(filter ContainerFilter) ([]Container, error)
+	// Inspect returns the detailed status of the container.
+	Inspect(id string) (ContainerInspectResult, error)
+}
+
+// ContainerCommandExecutor provides methods to run a command in the container.
+// TODO: Should we merge this with ContainerRuntime?
+type ContainerCommandExecutor interface {
+	// Exec executes a command in the container.
+	Exec(id string, cmd []string, streamOpts StreamOptions) error
+}
+
+// ContainerInspectResult contains the detailed config and status for a
+// container.
+type ContainerInspectResult struct {
+	// ID of the container.
+	ID string
+	// Config of the container.
+	Config ContainerConfig
+	// Status of the container.
+	// TODO: Re-examine ContainerStatus and consolidate the fields.
+	Status ContainerStatus
+}
+
+type ContainerConfig struct {
+	// Name of the container.
+	Name string
+	// Image to use.
+	Image ImageSpec
+	// RootFSPath is the path to the root filesystem. This field is optional.
+	// If not set, the runtime should create the root filesystem itself.
+	RootFSPath *string
```

Comment:
In that case, I'd compose `Image` object to contain either a `RootFsPath`
or `ImageName` with `Type`. `Type` is useful to specify docker vs appc.

On Tue, May 17, 2016 at 10:07 AM, Yu-Ju Hong <notifications@github.com>
wrote:

> In pkg/kubelet/container/interface.go
> <https://github.com/yujuhong/kubernetes/pull/1#discussion_r63562932>:
>
> > +	ID string
> > +	// Config of the container.
> > +	Config ContainerConfig
> > +	// Status of the container.
> > +	// TODO: Re-examine ContainerStatus and consolidate the fields.
> > +	Status ContainerStatus
> > +}
> > +
> > +type ContainerConfig struct {
> > +	// Name of the container.
> > +	Name string
> > +	// Image to use.
> > +	Image ImageSpec
> > +	// RootFSPath is the path to the root filesystem. This field is optional.
> > +	// If not set, the runtime should create the root filesystem itself.
> > +	RootFSPath *string
>
> Image may still be valuable to include so that it's clear what image is
> used to create the container.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/yujuhong/kubernetes/pull/1/files/ddd041c68ca94d95fa591dce703932fe5d577dd0#r63562932>
>